Add nullbits TIDBIT

jaygreco opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description
Add the nullbits TIDBIT to mainline QMK. Up to this point I had maintained a [separate repo](https://github.com/nullbitsco/tidbit) with the keymaps for this board. Creating a PR for this now that our common file cleanup has been merged and I have some time. 

Most keymaps are unremarkable. The few exceptions are ~~the 14seg keymap, which includes files for interfacing with an I2C 14-segment alphanumeric display, and~~ the snailmap_lite keymap, which is a numpad port of [this awesome keymap](https://github.com/qmk/qmk_firmware/tree/master/keyboards/nullbitsco/nibble/keymaps/snailmap) for the NIBBLE. 

All keymaps have been tested extensively on my dev boards and are working with the latest QMK.

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* N/A

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

jaygreco commented:
Checking in, I don't want this to go stale. Thanks!

drashna commented on keyboards/nullbitsco/tidbit/keymaps/14seg/HT16K33.h:
```diff
@@ -0,0 +1,94 @@
+/* Copyright 2021 Jay Greco
```
please rename this file to be lower cased.

drashna commented on keyboards/nullbitsco/tidbit/keymaps/14seg/keymap.c:
```diff
@@ -0,0 +1,196 @@
+};
+
+void matrix_init_user(void) {
+  matrix_init_remote_kb();
```
this should be called at the kb level, not here. 
```suggestion
```

drashna commented on keyboards/nullbitsco/tidbit/keymaps/14seg/keymap.c:
```diff
@@ -0,0 +1,196 @@
+}
+
+void matrix_scan_user(void) {
+  matrix_scan_remote_kb();
```
```suggestion
```

drashna commented on keyboards/nullbitsco/tidbit/keymaps/14seg/keymap.c:
```diff
@@ -0,0 +1,196 @@
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  process_record_remote_kb(keycode, record);
```
```suggestion

drashna commented on keyboards/nullbitsco/tidbit/keymaps/4rotary/config.h:
```diff
@@ -0,0 +1,19 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// Encoders are defined in order. 1: B2 & B3, 2: B4 & B5, 3: D0 & D1, 4: D2 & D3
```
```suggestion
#pragma once

// Encoders are defined in order. 1: B2 & B3, 2: B4 & B5, 3: D0 & D1, 4: D2 & D3
```

drashna commented on keyboards/nullbitsco/tidbit/keymaps/4rotary/keymap.c:
```diff
@@ -0,0 +1,68 @@
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  process_record_remote_kb(keycode, record);
```
```suggestion
```

drashna commented on keyboards/nullbitsco/tidbit/keymaps/4rotary/keymap.c:
```diff
@@ -0,0 +1,68 @@
+}
+
+void matrix_init_user(void) {
+  matrix_init_remote_kb();
```
```suggestion
```

drashna commented on keyboards/nullbitsco/tidbit/keymaps/4rotary/keymap.c:
```diff
@@ -0,0 +1,68 @@
+  set_bitc_LED(LED_OFF);
+}
+
+void matrix_scan_user(void) {
+  matrix_scan_remote_kb();
+}
```
```suggestion
```

drashna commented on keyboards/nullbitsco/tidbit/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,68 @@
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  process_record_remote_kb(keycode, record);
```
```suggestion
```

drashna commented on keyboards/nullbitsco/tidbit/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,68 @@
+}
+
+void matrix_init_user(void) {
+  matrix_init_remote_kb();
```
```suggestion
```

drashna commented on keyboards/nullbitsco/tidbit/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,68 @@
+  set_bitc_LED(LED_OFF);
+}
+
+void matrix_scan_user(void) {
+  matrix_scan_remote_kb();
+}
```
```suggestion
```

drashna commented on keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c:
```diff
@@ -0,0 +1,149 @@
+};
+
+
+#ifdef OLED_ENABLE
```
please move the oled code to the keyboard c file, and use the oled_task_kb instead

jaygreco replied:
This was the only change I wasn't sure about. There is more than one keymap with OLED functionality -- would moving the OLED code to the keyboard file mean it applies to all keymaps with the OLED enabled?

drashna replied:
Moving the code to the keyboard's c file allows any of the keymaps to have that set as the default, but can allow for the display to be overridden by users.   It also allows the board to have a default config for the QMK Configurator.

jaygreco replied:
Ah, that makes sense. Changed in the latest revision. Thanks!

drashna commented on keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c:
```diff
@@ -0,0 +1,149 @@
+void oled_task_user(void) { render_logo(); }
+#endif
+
+void matrix_init_user(void) { 
+    matrix_init_remote_kb();
+}
+
+void matrix_scan_user(void) {
+    matrix_scan_remote_kb();
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    process_record_remote_kb(keycode, record);
```
```suggestion
bool process_record_user(uint16_t keycode, keyrecord_t *record) {
```

drashna commented on keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    switch (keycode) {
+        case PROG:
```
```suggestion
        case RESET:
```

drashna commented on keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c:
```diff
@@ -0,0 +1,149 @@
+                oled_off();
+                bootloader_jump();  // jump to bootloader
+            }
+            break;
```
```suggestion
            return false;
```

drashna commented on keyboards/nullbitsco/tidbit/keymaps/14seg/config.h:
```diff
@@ -0,0 +1,20 @@
+
+
+// Encoders are defined in order. 1: B2 & B3, 2: B4 & B5, 3: D0 & D1, 4: D2 & D3
+#define ENCODERS_PAD_A { B2, B4 }
```
these should be defined at the keyboard level

jaygreco commented:
I added two commits: the first removes keymaps that weren't worth including after thinking about it. The second implements the PR feedback. I also cleaned up the file formatting and indentation to adhere to the QMK coding convention.

jaygreco commented:
Thanks @drashna for the PR feedback!

jaygreco commented:
@drashna great advice on moving a lot of this stuff to the keyboard file. I pulled out more beyond what you'd recommended in the latest commit. Thanks for the feedback!

drashna commented on keyboards/nullbitsco/tidbit/keymaps/snailmap_lite/keymap.c:
```diff
@@ -0,0 +1,294 @@
+    uint8_t cursorposition_cur = 13;
+    uint8_t cursorposition_max = 14;
+
+    char wpm_counter[4];
+    wpm_counter[3] = '\0';
+    wpm_counter[2] = '0' + current_wpm % 10;
+    wpm_counter[1] = '0' + (current_wpm / 10) % 10;
+    wpm_counter[0] = '0' + (current_wpm / 100) % 10;
+    oled_set_cursor(0, cursorposition_cur);
+    oled_write(wpm_counter, false);
```
```suggestion
    oled_set_cursor(0, cursorposition_cur);
    oled_write(get_u8_str(current_wpm), false);
```

drashna commented on keyboards/nullbitsco/tidbit/rules.mk:
```diff
@@ -0,0 +1,28 @@
+LTO_ENABLE = yes            # Link-time optimization
+
+# Project specific files
+SRC += i2c_master.c \
+    common/bitc_led.c \
+    common/remote_kb.c \
+    uart.c
```
```suggestion
SRC += common/bitc_led.c \
    common/remote_kb.c
QUANTUM_LIB_SRC += i2c_master.c \
    uart.c
```

jaygreco commented:
@drashna Feedback incorporated. I also removed the WPM fudge factor from the same keymap at the same time.

fauxpark commented on keyboards/nullbitsco/tidbit/rules.mk:
```diff
@@ -0,0 +1,28 @@
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+AUDIO_ENABLE = no           # Audio output on port C6
+ENCODER_ENABLE = yes        # Use rotary encoder
+LTO_ENABLE = yes            # Link-time optimization
```
```suggestion
BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = no            # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
ENCODER_ENABLE = yes
LTO_ENABLE = yes
```

jaygreco commented:
Hi @drashna and @fauxpark, apologies for the ping! I'm just checking in on this. I'm hoping to get it merged before the breaking changes push this month. Thanks!

noroadsleft commented:
Thanks!
